design-proposals/klip-27-enhanced-udf-configuration-options.md

```diff
@@ -0,0 +1,80 @@
+# KLIP 27 - Enhanced UDF Configuration Options
+
+**Author**: Hans-Peter Grahsl (@hpgrahsl) | 
+**Release Target**: ??? | 
+**Status**: _Approved_ | 
+**Discussion**: [GitHub PR](https://github.com/confluentinc/ksql/pull/5269)
+
+**tl;dr:** This KLIP suggests to introduce a new configuration option in order to externally define UDF properties, that can be shared among a group / family of (custom) UDFs which logically belong together. At the same time, support for the `_global_` scope of configuration settings should get dropped.
+
+## Motivation and background
+
+Currently, ksqlDB supports two different ways to configure UDFs based on entries defined in `ksql-server.properties`:
+
+* either the configuration properties are directly bound to a single, specific UDF class and applicable to potentially existing overloaded methods - e.g. `ksql.functions.<myudfname>.<mysetting>=foo`
+
+* or we can have properties defined in a special scope, making them globally accessible - e.g. `ksql.functions._global_.<mysetting>=1234`
+
+The latter approach means that properties are readable by any(!) other UDF that gets loaded during bootstrap. While this may or may not be an issue depending on the type of properties, in general, it can be considered less than ideal. For instance, this leads to security concerns for sensitive configuration properties such as secrets / tokens / passwords, which is why the `_global_` scope should get dropped.
+
+## What is in scope
+
+This KLIP aims to allow for multiple, yet specifically defined UDFs, to share configuration properties other than "misusing" the global properties scope.
```

Comment:
Will the scope cover UDAFs and UDTFs too?

Reply:
I would say yes of course, at least I don't see any reasonable argument why this configuration options change should only affect UDFs. Will adapt the text to make that clear.

Reply:
Can we call out UDTFs and UDAFs in the scope to be clear please?